docs/yaml/builtins/meson.yaml

```diff
@@ -32,6 +32,17 @@ methods:
       scripts, but for subproject scripts they have the path to the root of the
       subproject appended, usually `subprojects/<subproject-name>`.
 
+      *(since 1.4.0)* The `MESONREWRITE` environment variable contains the path
+      to the rewrite command that corresponds to the `meson` executable that
+      was used to configure the build. (This might be a different path than the
+      first executable found in `PATH`.) It can be used to remove or replace
+      any `run_command` that depends on the revision control system from the
+      build configuration. Note that the value will contain many parts. For
+      example, it may be `python3 /path/to/meson.py introspect`. The user is
+      responsible for splitting the string to an array if needed by splitting
+      lexically like a UNIX shell would. If your script uses Python,
+      `shlex.split()` is the easiest correct way to do this.
```

Comment:
Hmm, you say this and yet you break that exact logic in your first commit -- which has the commit message:
```
Use mesonlib.join_args consistently 
```

and switches away from consistently using shlex, migrating over to a meson-specific function that behaves differently on Windows vs. Unix and is intended to format command lines in a cmd.exe-compatible manner.

Reply:
That text was copied from the documentation of `MESONINTROSPECT`. I see that the first command indeed breaks some things on Windows, but I wonder how they could have worked before. In `minstall.py`, `MESONINTROSPECT` was defined using `' '.join(shlex.quote(...) ...)`, whereas in `backend/backends.py` that same variable ends up being defined using `join_args`...

I can drop the first commit, but then please tell me what to use in the final commit: `shlex.quote`, or `join_args`.

Reply:
This discrepancy started with 522392e (#8305), where the code in `backend/backends.py` changed from `' '.join(shlex.quote(...) ...)` to `parse_args`. There is an inconsistency here and I don't which (`backend/backends.py` or `minstall.py`) is 'correct' here. I guess it depends on whether we expect all scripts to run from Python (`shlex`), or from the applicable shell (for which `shlex` doesn't do the right thing on Windows).

Maybe @xclaesse (author of #8305) can weigh in?

Reply:
That code was clearly a subtle regression at the time (years ago!) and went unnoticed due to a combination of:
- the feature itself is somewhat obscure
- on Linux, the development platform of the contributor, it behaves identically. :D

I'm of the opinion that you cannot safely parse windows formatted command lines but posix ones are trivial and even if they aren't run in cmd.exe using posix formatting, posix formatting is the only sane way to communicate them back and forth.

Reply:
I share this viewpoint. I'll rework the first commit to only address the regression (and see if I can add a test). This would fix #12148. In addition, I'll open an issue suggesting we drop the Windows quoting madness altogether: you can't have quoting that is easily usable in Python and `cmd.exe` at the same time. That may be a shortcoming of Python, but it is not something that can be fixed in meson unless we suggest people use `mesonbuild.quote` and `mesonbuild.split` instead of `shlex` functions.

Reply:
I've re-focused the current pull request on `MESONREWRITE` and moved the `join_args` part out into #12807.